### ISSUE TITLE
3.9.2: chunkserver: random chunk errors on restart

### ISSUE THREAD

**Issue by onlyjob** [lizardfs/lizardfs#324]
Chunkserver upgraded from 2.6.0 and configured with `CREATE_NEW_CHUNKS_IN_MOOSEFS_FORMAT = 0` exhibit random chunk errors (visible in CGI/Disks) on every restart. Below are chunkserver log fragments from 3 consecutive restarts:

~~~~
Nov 02 08:44:57 debmain systemd[1]: Stopping LizardFS chunkserver daemon...
Nov 02 08:44:59 debmain systemd[1]: Starting LizardFS chunkserver daemon...
Nov 02 08:44:59 debmain systemd[1]: Started LizardFS chunkserver daemon.
Nov 02 08:44:59 debmain mfschunkserver[23251]: mfschunkserver daemon initialized properly
Nov 02 08:45:11 debmain mfschunkserver[23251]: mfschunkserver[23251]: scanning folder /var/lib/lizardfs/05/: complete (12s)
Nov 02 08:45:22 debmain mfschunkserver[23251]: set_chunk_version: file:/var/lib/lizardfs/02/chunksC5/chunk_xor_1_of_2_0000000009C52BA5_00000001.liz - rename error: ENOENT (No such file or directory)
Nov 02 08:45:22 debmain mfschunkserver[23251]: set_chunk_version: file:/var/lib/lizardfs/03/chunksC5/chunk_xor_1_of_2_0000000009C526D7_00000001.liz - rename error: ENOENT (No such file or directory)
Nov 02 08:45:22 debmain mfschunkserver[23251]: delete_chunk: file:/var/lib/lizardfs/02/chunksC5/chunk_xor_1_of_2_0000000009C52BA5_00000001.liz - unlink error: ENOENT (No such file or directory)
Nov 02 08:45:22 debmain mfschunkserver[23251]: delete_chunk: file:/var/lib/lizardfs/03/chunksC5/chunk_xor_1_of_2_0000000009C526D7_00000001.liz - unlink error: ENOENT (No such file or directory)
Nov 02 08:45:22 debmain mfschunkserver[23251]: replication error: failed to set chunk's version (IO error)
Nov 02 08:45:22 debmain mfschunkserver[23251]: replication error: failed to set chunk's version (IO error)
Nov 02 08:45:25 debmain mfschunkserver[23251]: set_chunk_version: file:/var/lib/lizardfs/04/chunksFA/chunk_0000000002FA4F2A_00000001.liz - rename error: ENOENT (No such file or directory)
Nov 02 08:45:25 debmain mfschunkserver[23251]: delete_chunk: file:/var/lib/lizardfs/04/chunksFA/chunk_0000000002FA4F2A_00000001.liz - unlink error: ENOENT (No such file or directory)
Nov 02 08:45:25 debmain mfschunkserver[23251]: replication error: failed to set chunk's version (IO error)
Nov 02 08:45:25 debmain mfschunkserver[23251]: mfschunkserver[23251]: set_chunk_version: file:/var/lib/lizardfs/04/chunksFA/chunk_0000000002FA4F2A_00000001.liz - rename error: ENOENT (No such file or direct
Nov 02 08:45:25 debmain mfschunkserver[23251]: mfschunkserver[23251]: delete_chunk: file:/var/lib/lizardfs/04/chunksFA/chunk_0000000002FA4F2A_00000001.liz - unlink error: ENOENT (No such file or directory)
Nov 02 08:45:25 debmain mfschunkserver[23251]: mfschunkserver[23251]: replication error: failed to set chunk's version (IO error)
~~~~

~~~~
Nov 02 08:46:08 debmain systemd[1]: Stopping LizardFS chunkserver daemon...
Nov 02 08:46:08 debmain mfschunkserver[23251]: terminate signal received
Nov 02 08:46:10 debmain systemd[1]: Starting LizardFS chunkserver daemon...
Nov 02 08:46:10 debmain systemd[1]: Started LizardFS chunkserver daemon.
Nov 02 08:46:10 debmain mfschunkserver[23878]: mfschunkserver daemon initialized properly
Nov 02 08:46:10 debmain mfschunkserver[23878]: [ OK ] mfschunkserver daemon initialized properly
Nov 02 08:46:10 debmain mfschunkserver[23878]: connected to Master
Nov 02 08:46:32 debmain mfschunkserver[23878]: set_chunk_version: file:/var/lib/lizardfs/04/chunksC3/chunk_xor_1_of_2_0000000009C3D033_00000001.liz - rename error: ENOENT (No such file or directory)
Nov 02 08:46:32 debmain mfschunkserver[23878]: delete_chunk: file:/var/lib/lizardfs/04/chunksC3/chunk_xor_1_of_2_0000000009C3D033_00000001.liz - unlink error: ENOENT (No such file or directory)
Nov 02 08:46:32 debmain mfschunkserver[23878]: replication error: failed to set chunk's version (IO error)
Nov 02 08:46:32 debmain mfschunkserver[23878]: set_chunk_version: file:/var/lib/lizardfs/05/chunks9E/chunk_00000000009EEC19_00000001.liz - rename error: ENOENT (No such file or directory)
Nov 02 08:46:32 debmain mfschunkserver[23878]: delete_chunk: file:/var/lib/lizardfs/05/chunks9E/chunk_00000000009EEC19_00000001.liz - unlink error: ENOENT (No such file or directory)
Nov 02 08:46:32 debmain mfschunkserver[23878]: replication error: failed to set chunk's version (IO error)
Nov 02 08:46:32 debmain mfschunkserver[23878]: set_chunk_version: file:/var/lib/lizardfs/05/chunks9E/chunk_00000000009EE469_00000001.liz - rename error: ENOENT (No such file or directory)
Nov 02 08:46:32 debmain mfschunkserver[23878]: delete_chunk: file:/var/lib/lizardfs/05/chunks9E/chunk_00000000009EE469_00000001.liz - unlink error: ENOENT (No such file or directory)
Nov 02 08:46:32 debmain mfschunkserver[23878]: replication error: failed to set chunk's version (IO error)
Nov 02 08:46:33 debmain mfschunkserver[23878]: mfschunkserver[23878]: scanning folder /var/lib/lizardfs/05/: complete (23s)
~~~~

~~~~
Nov 02 08:47:05 debmain systemd[1]: Stopping LizardFS chunkserver daemon...
Nov 02 08:47:05 debmain mfschunkserver[23878]: terminate signal received
Nov 02 08:47:07 debmain mfschunkserver[24371]: mfschunkserver daemon initialized properly
Nov 02 08:47:07 debmain mfschunkserver[24371]: connected to Master
Nov 02 08:47:25 debmain mfschunkserver[24371]: replication error: failed to create chunk (Chunk already exists)
Nov 02 08:47:25 debmain mfschunkserver[24371]: replication error: failed to create chunk (Chunk already exists)
Nov 02 08:47:29 debmain mfschunkserver[24371]: set_chunk_version: file:/var/lib/lizardfs/05/chunksC0/chunk_0000000001C0191A_00000003.liz - rename error: ENOENT (No such file or directory)
Nov 02 08:47:29 debmain mfschunkserver[24371]: delete_chunk: file:/var/lib/lizardfs/05/chunksC0/chunk_0000000001C0191A_00000003.liz - unlink error: ENOENT (No such file or directory)
Nov 02 08:47:29 debmain mfschunkserver[24371]: replication error: failed to set chunk's version (IO error)
~~~~

When `CREATE_NEW_CHUNKS_IN_MOOSEFS_FORMAT = 1` (reverted to default) chunkserver do not throw chunk errors upon restart.

**Comment by onlyjob** [comment]
This bug may be responsible for loss of _xor2_ chunks on restart... Under some circumstances reboot produces missing _xor2_ chunks...

**Comment by onlyjob** [comment]
Chunkservers on latest "master" branch mostly complain on start like

    write_block_to_chunk: file:/var/lib/lizardfs/04/chunksFD/chunk_0000000007FDBEDB_00000001.liz - crc write error: EBADF (Bad file descriptor)
    hdd_io_end: refcount = 0 - This should never happen!

But the outcome is the same -- spurious errors in CGI/Disks...

**Comment by onlyjob** [comment]
Probably related to #349.

**Comment by onlyjob** [comment]
No longer observe this on 3.9.4 -- likely fixed by 069139a. Closing...
